### DISCUSSION TITLE
Is this the correct way to call register?

### DISCUSSION THREAD

**Discussion by eduleite** [react-hook-form/react-hook-form#2742] [Q&A] [answered]
Guys,

Just found a situation inside my project that required me to manually register a field.
The sandbox below show the exact situation that happened in my project.
I Have a form that is updated with values from a database using the reset call.
My field is manually registered inside a useEffect that runs once. If I submit the form without editing the field and calling the setValue, the values object inside the function called by handleSubmit is empty.
But if I instead call register in every render (or whenever register is updated) the values object has the value from the manually registered field.

Is it correct to call the register in every render? Or it is a bug? All examples that I see calls register once, but in practice it is not happening this way. Even the shouldDirty is not working properly in this situation. It only works when I register in every render.

EDIT: Just saw that source code of the Controller component and it apears that it calls register on every render.

https://codesandbox.io/s/heuristic-elion-xusuw?file=/src/App.tsx

**Comment by bluebill1049** [comment, marked as answer]
hey @eduleite we don't' call register each render at Controller, can you share the steps to see the issue?

**Reply by eduleite** [reply]
@bluebill1049, just click the submit button, and see the console log that the values object passed from handleSubmit is empty.
If I add [register] in the dependency list of the useEffect call that registers, it will work as expected.

**Reply by eduleite** [reply]
When I mean every render, I mean whenever register changes. It is in the useCallback dependency list inside the Controller source.

**Reply by bluebill1049** [reply]
this usage is probably not going to be supported, you are mixing controlled with uncontrolled, if you need to use a controlled input, perhaps just use controller?

**Reply by eduleite** [reply]
I have a field in my form that I have to listen to in other to update other part of the form. Whenever the user types the 4th letter, I hava to call a web service to validate that field and fill other 2 fields.
I tried using the Controller in this field the first time, but the useWatch was not being updated when the user was typing... I will try it again, I may be doing something wrong.

**Reply by eduleite** [reply]
Well... switched to Controller and the useWatch is being updated... back to the real code to see what am I doing wrong there.

**Reply by bluebill1049** [reply]
yea, keep me posted with it.

**Reply by eduleite** [reply]
Yeah, nothing to see here... I was doing something wrong, maybe a property name being misspelled

**Reply by eduleite** [reply]
Is there a pattern to help to reduce this mispelling object property names? As time goes on, all of my troubles end up being messing with the object property names.

**Reply by bluebill1049** [reply]
we are introducing more run-time checks for the incorrect names now.

**Reply by bluebill1049** [reply]
still please let me which part are the pain points tho.

**Reply by eduleite** [reply]
My only gripe with the library for now is really when you mess with the property names. Just migrated my first big form (20 fields) and it was the main problem I had. A way to check in development mode and runtime if we are making this mistakes will help a lot.
Just fixed my form to use the controller and the error is gone. It was really something on my part that I was doing wrong.

**Reply by bluebill1049** [reply]
will include more name-checking, thanks for the feedback.
